File: php/src/Game.php
<?php

namespace TicTacToe;

use Exception;

class Game
{
    /** @var string */
    private $_lastSymbol = ' ';
    
    /** @var Board */
    private $_board;

    public function __construct()
    {
        $this->_board = new Board();
    }

    /**
     * @throws Exception
     */
    public function play(string $symbol, int $x, int $y): void
    {
        //if first move
        if ($this->_lastSymbol == ' ') {
            //if player is X
            if ($symbol == 'O') {
                throw new Exception("Invalid first player");
            }
        }
        //if not first move but player repeated
        else if ($symbol == $this->_lastSymbol) {
            throw new Exception("Invalid next player");
        }
        //if not first move but play on an already played tile
        else if ($this->_board->tileAt($x, $y)->symbol != ' ') {
            throw new Exception("Invalid position");
        }

        // update game state
        $this->_lastSymbol = $symbol;
        $this->_board->addTileAt($symbol, $x, $y);
    }

    public function winner(): string
    {
        //if the positions in first row are taken
        if ($this->_board->tileAt(0, 0)->symbol != ' ' &&
            $this->_board->tileAt(0, 1)->symbol != ' ' &&
            $this->_board->tileAt(0, 2)->symbol != ' ') {
            //if first row is full with same symbol
            if ($this->_board->tileAt(0, 0)->symbol ==
                $this->_board->tileAt(0, 1)->symbol &&
                $this->_board->tileAt(0, 2)->symbol == $this->_board->tileAt(0, 1)->symbol) {
                return $this->_board->tileAt(0, 0)->symbol;
            }
        }

        //if the positions in first row are taken
        if ($this->_board->tileAt(1, 0)->symbol != ' ' &&
            $this->_board->tileAt(1, 1)->symbol != ' ' &&
            $this->_board->tileAt(1, 2)->symbol != ' ') {
            //if middle row is full with same symbol
            if ($this->_board->tileAt(1, 0)->symbol ==
                $this->_board->tileAt(1, 1)->symbol &&
                $this->_board->tileAt(1, 2)->symbol ==
                $this->_board->tileAt(1, 1)->symbol) {
                return $this->_board->tileAt(1, 0)->symbol;
            }
        }

        //if the positions in first row are taken
        if ($this->_board->tileAt(2, 0)->symbol != ' ' &&
            $this->_board->tileAt(2, 1)->symbol != ' ' &&
            $this->_board->tileAt(2, 2)->symbol != ' ') {
            //if middle row is full with same symbol
            if ($this->_board->tileAt(2, 0)->symbol ==
                $this->_board->tileAt(2, 1)->symbol &&
                $this->_board->tileAt(2, 2)->symbol ==
                $this->_board->tileAt(2, 1)->symbol) {
                return $this->_board->tileAt(2, 0)->symbol;
            }
        }

        return ' ';
    }
}

File: python/Game.py
from Board import Board


class Game(object):

    def __init__(self):
        self._lastSymbol = ' '
        self._board = Board()

    def play(self, symbol, x, y):
        # if first move
        if self._lastSymbol == ' ':
            # if player is X
            if symbol == 'O':
                raise Exception('Invalid first player')
        # if not first move but player repeated
        elif symbol == self._lastSymbol:
            raise Exception('Invalid next player')
        # if not first move but play on an already played tile
        elif self._board.TileAt(x, y).Symbol != ' ':
            raise Exception('Invalid position')

        # update game state
        self._lastSymbol = symbol
        self._board.AddTileAt(symbol, x, y)

    def winner(self):
        # if the positions in first row are taken
        if self._board.TileAt(0, 0).Symbol != ' ' \
                and self._board.TileAt(0, 1).Symbol != ' ' \
                and self._board.TileAt(0, 2).Symbol != ' ':
            # if first row is full with same symbol
            if self._board.TileAt(0, 0).Symbol == self._board.TileAt(0, 1).Symbol \
                    and self._board.TileAt(0, 2).Symbol == self._board.TileAt(0, 1).Symbol:
                return self._board.TileAt(0, 0).Symbol

        # if the positions in first row are taken
        if self._board.TileAt(1, 0).Symbol != ' ' \
                and self._board.TileAt(1, 1).Symbol != ' ' \
                and self._board.TileAt(1, 2).Symbol != ' ':
            # if first row is full with same symbol
            if self._board.TileAt(1, 0).Symbol == self._board.TileAt(1, 1).Symbol \
                    and self._board.TileAt(1, 2).Symbol == self._board.TileAt(1, 1).Symbol:
                return self._board.TileAt(1, 0).Symbol

        # if the positions in first row are taken
        if self._board.TileAt(2, 0).Symbol != ' ' \
                and self._board.TileAt(2, 1).Symbol != ' ' \
                and self._board.TileAt(2, 2).Symbol != ' ':
            # if first row is full with same symbol
            if self._board.TileAt(2, 0).Symbol == self._board.TileAt(2, 1).Symbol \
                    and self._board.TileAt(2, 2).Symbol == self._board.TileAt(2, 1).Symbol:
                return self._board.TileAt(2, 0).Symbol

        return ' '
